### DISCUSSION TITLE
socketio websocket server with tls/ssl

### DISCUSSION THREAD

**Discussion by almaz1213** [miguelgrinberg/python-socketio#1090] [Q&A]
Hi guys,

as i understand from the discussions it's impossible to use socketio.AsyncServer() to serve wss connection from browsers without external web server.

Do you have a solution to wrap it someting like
```
import ssl, socketio, http.server
sio = socketio.AsyncServer()
ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
ssl_context.load_cert_chain('cert.csr', 'key.pem')
sio = ssl_context.wrap_socket(sio, server_side=True)

<and launch http.server>
```

It is just for secure transport from-to browser and the app must work .
Certificate is self-signed.

Thanks in advance

**Comment by miguelgrinberg** [comment]
I'm really confused about what you are saying and don't know what in the discussions here made you think that.

python-socketio is just a web framework, it does not deal with serving or encrypting content, that is left for web servers to do. All the web servers that you can use with python-socketio have support for TLS/SSL, so what is the issue? What made you think this is impossible? Just look at the documentation for the [list of web servers](https://python-socketio.readthedocs.io/en/latest/server.html#deployment-strategies) that you can use, pick one, then learn how to run that server with TLS/SSL. Or run the web server w/o encryption and put Nginx in front to handle that encrypted connections.

**Comment by almaz1213** [comment]
@miguelgrinberg got your opinion.

Anybody else who understand that I need not launch external web server and I need use python socketio with self signed ssl/tls cert just in the server app can help me?

Could someone provide me a solution/example using e.g http.server module right in the server app that uses socketio, please?

The server app has no domain name. It must be available via IP address only and each time when the IP address of the node is changed, it generates new self signed certs for that IP when it restarts.

The websocket server must be connectable from browsers.

Thanks in advance

**Reply by miguelgrinberg** [reply]
@almaz1213 Have you seen the little label that appears next to my name, the one that reads "maintainer"? I've made this and I'm giving you facts, not opinion.

This package does not include a web server, the web server has to be added. I have never attempted to use the `http.server` module, but you are welcome to give it a try if you want. It is going to require some work, however, for the following reasons:

- `http.server` is not asyncio compatible, so it will never be able to work alongside the `socketio.AsyncServer` class you indicated that you want to use. You may be able to make it work with `socketio.Server`.
- `http.server` does not implement WebSocket, so if you make it work with this package, Socket.IO will be limited to use long-polling.
- `http.server` does not implement the WSGI or ASGI protocols, so to use it with this package, you will have to build your own integration/glue layer.

Hope this helps.

**Reply by almaz1213** [reply]
probably i should build own solution using asgiref...

**Reply by miguelgrinberg** [reply]
asgiref is not a webserver either, you will still need to add something like uvicorn.

**Comment by almaz1213** [comment]
i didn't say that asgiref is a web server. i said "using asgiref".
what about aiohttp?

**Reply by miguelgrinberg** [reply]
https://python-socketio.readthedocs.io/en/latest/server.html#aiohttp

**Comment by almaz1213** [comment]
> Have you seen the little label that appears next to my name, the one that reads "maintainer"?

thanks. and thanks to engine.io
